src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreePostingsWriter.java

```diff
@@ -0,0 +1,198 @@
+    {
+        checkArgument(!pathToRoot.containsInt(leafNodeID));
+        checkArgument(pathToRoot.isEmpty() || leafNodeID > pathToRoot.get(pathToRoot.size() - 1));
+
+        leafOffsetToNodeID.put(leafBlockFP, leafNodeID);
+        for (int i = 0; i < pathToRoot.size(); i++)
+        {
+            int level = i + 1;
+            if (isLevelEligibleForPostingList(level))
+            {
+                int nodeID = pathToRoot.get(i);
+                nodeToChildLeaves.put(nodeID, leafNodeID);
+            }
+        }
+    }
+
+    public long finish(IndexOutput out) throws IOException
+    {
+        checkState(postings.size() == leafOffsetToNodeID.size(),
+                   "Expected equal number of postings lists (%s) and leaf offsets (%s).",
+                   postings.size(), leafOffsetToNodeID.size());
+
+        try (PostingsWriter postingsWriter = new PostingsWriter(out))
+        {
+            Iterator<PackedLongValues> postingsIterator = postings.iterator();
+            Map<Integer, PackedLongValues> leafToPostings = new HashMap<>();
+            leafOffsetToNodeID.forEach((fp, nodeID) -> leafToPostings.put(nodeID, postingsIterator.next()));
+
+            long postingsRamBytesUsed = postings.stream()
+                                                .mapToLong(PackedLongValues::ramBytesUsed)
+                                                .sum();
+
+            List<Integer> internalNodeIDs = nodeToChildLeaves.keySet()
+                                                             .stream()
+                                                             .filter(i -> nodeToChildLeaves.get(i).size() >= MINIMUM_POSTINGS_LEAVES)
+                                                             .collect(Collectors.toList());
+
+            Collection<Integer> leafNodeIDs = leafOffsetToNodeID.values();
+
+            logger.debug(indexContext.logMessage("Writing posting lists for {} internal and {} leaf balanced tree nodes. Leaf postings memory usage: {}."),
+                         internalNodeIDs.size(),
+                         leafNodeIDs.size(),
+                         FBUtilities.prettyPrintMemory(postingsRamBytesUsed));
+
+            long startFP = out.getFilePointer();
+            Stopwatch flushTime = Stopwatch.createStarted();
+            TreeMap<Integer, Long> nodeIDToPostingsFilePointer = new TreeMap<>();
+            for (int nodeID : Iterables.concat(internalNodeIDs, leafNodeIDs))
+            {
+                Collection<Integer> leaves = nodeToChildLeaves.get(nodeID);
+
+                if (leaves.isEmpty())
+                {
+                    leaves = Collections.singletonList(nodeID);
+                    numLeafPostings++;
+                }
+                else
+                {
+                    numNonLeafPostings++;
+                }
+
+                PriorityQueue<PeekablePostingList> postingLists = new PriorityQueue<>(100, Comparator.comparingLong(PeekablePostingList::peek));
```

Comment:
What if instead of creating this w/ an arbitrary starting size of 100 on every iteration, we create the `PriorityQueue` outside the loop w/ a starting size of `MINIMUM_POSTINGS_LEAVES` and just `clear()` it at the end of every iteration?